### DISCUSSION TITLE
Copy subject key identifier to authority key identifier

### DISCUSSION THREAD

**Discussion by kgoldman** [openssl/openssl#26026] [Q&A]
I have a CA certificate and user certificate as X509 structures. I would like to programmatically copy the CA subject key identifier to the user authority key identifier.

I have a sense that X509_get_ext_d2i(, NID_subject_key_identifier, ) can get the subject key identifier as an ASN1_OCTET STRING.  Is that correct?

How would I add that to the user cert, perhaps with NID_authority_key_identifier?

**Comment by nicferrier** [comment]
Did you get an answer here @kgoldman ? I want to do this too and documentation seems sparse.

**Reply by kgoldman** [reply]
I did not get an answer. I don't have anything that is currently available as open source.

**Reply by dndfrink-ibm** [reply]
This snippet might be helpful:

```
    X509* subjectCertificate;
    X509* issuerCertificate;
    /* initialize subject/issuer certs, set SKI on issuer cert with similar methods to below) */
    X509V3_CTX ctx;
    X509V3_set_ctx_nodb(&ctx);
    X509V3_set_ctx(&ctx, issuerCertificate, subjectCertificate, NULL, NULL, 0);

    X509_EXTENSION* ext = X509V3_EXT_conf_nid(NULL, &ctx, NID_authority_key_identifier, "keyid:always");
    if(ext)
    {
        X509_add_ext(subjectCertificate, ext, -1);
    }
```

**Reply by dndfrink-ibm** [reply]
the key is that you need to set both the issuer and subject certificates in the context for adding the AKI to work properly. For other extensions (such as setting the SKI), only setting the subject certificate in the context would be necessary (i.e., a self-signed/parent certificate would have no issuer, but it still needs an SKI).
